modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsWriter.java

```diff
@@ -0,0 +1,517 @@
+
+        doWrite(TASK, () -> {
+            int size = 24 + 1 + 4 + 8 + 8 + 4;
+
+            if (needWriteStr)
+                size += 4 + strBytes.length;
+
+            return size;
+        }, buf -> {
+            writeIgniteUuid(buf, sesId);
+            buf.put(needWriteStr ? (byte)1 : 0);
+            buf.putInt(taskName.hashCode());
+
+            if (needWriteStr) {
+                buf.putInt(strBytes.length);
+                buf.put(strBytes);
+            }
+
+            buf.putLong(startTime);
+            buf.putLong(duration);
+            buf.putInt(affPartId);
+        });
+    }
+
+    /**
+     * @param sesId Session id.
+     * @param queuedTime Time job spent on waiting queue.
+     * @param startTime Start time in milliseconds.
+     * @param duration Job execution time.
+     * @param timedOut {@code True} if job is timed out.
+     */
+    public void job(IgniteUuid sesId, long queuedTime, long startTime, long duration, boolean timedOut) {
+        doWrite(JOB,
+            () -> 24 + 8 + 8 + 8 + 1,
+            buf -> {
+                writeIgniteUuid(buf, sesId);
+                buf.putLong(queuedTime);
+                buf.putLong(startTime);
+                buf.putLong(duration);
+                buf.put(timedOut ? (byte)1 : 0);
+            });
+    }
+
+    /**
+     * @param op Operation type.
+     * @param sizeSupplier Record size supplier.
+     * @param writer Record writer.
+     */
+    private void doWrite(OperationType op, IntSupplier sizeSupplier, Consumer<ByteBuffer> writer) {
+        FileWriter fileWriter = this.fileWriter;
+
+        // Writer stopping.
+        if (fileWriter == null)
+            return;
+
+        int size = sizeSupplier.getAsInt() + /*type*/ 1;
+
+        SegmentedRingByteBuffer ringBuf = ringByteBuffer;
+
+        // Starting.
+        if (ringBuf == null)
+            return;
+
+        SegmentedRingByteBuffer.WriteSegment seg = ringBuf.offer(size);
+
+        if (seg == null) {
+            if (smallBufLogged.compareAndSet(false, true)) {
+                log.warning("The performance statistics in-memory buffer size is too small. Some operations " +
+                    "will not be logged.");
+            }
+
+            return;
+        }
+
+        // Ring buffer closed (writer stopping) or maximum size reached.
+        if (seg.buffer() == null) {
+            seg.release();
+
+            if (!fileWriter.isCancelled() && stopByMaxSize.compareAndSet(false, true)) {
+                stopStatistics();
+
+                log.warning("The performance statistics file maximum size is reached. " +
+                    "Performance statistics collecting will be stopped.");
+            }
+
+            return;
+        }
+
+        ByteBuffer buf = seg.buffer();
+
+        buf.put((byte)op.ordinal());
+
+        writer.accept(buf);
+
+        seg.release();
+
+        int readySize = readyForFlushSize.addAndGet(size);
```

Comment:
We have a race here.

1. FileWriter - writes some data and decrease `readyForFlushSize`.
2. Other threads executes `doWrite` and constantly increase `readySize` so it's become bigger than `DFLT_FLUSH_SIZE`. This leads that those threads will have contention of `wakeUp` synchronization.

Reply:
I think we should do the following:

1. Each time `readyForFlushSize` cross `DFLT_FLUSH_SIZE` we should clear it to zero.
2. Each time we see zero as a result of `updateAndGet` we should notify `FileWriter` to be sure it start flushing.
3. `FileWriter` just flush all the data it have for the moment(slightly more then `DFLT_FLUSH_SIZE` because of concurrent writes.

We shouldn't touch `readyForFlushSize` in the `FileWriter`.

Reply:
Done.